rust/src/lib.rs

```diff
@@ -106,430 +139,387 @@ pub extern "C" fn zcashlc_clear_last_error() {
     ffi_helpers::error_handling::clear_last_error()
 }
 
-/// Sets up the internal structure of the data database.
-#[no_mangle]
-pub extern "C" fn zcashlc_init_data_database(
-    db_data: *const u8,
-    db_data_len: usize,
-    network_id: u32,
-) -> i32 {
-    let res = catch_panic(|| {
-        let network = parse_network(network_id)?;
-        let db_data = Path::new(OsStr::from_bytes(unsafe {
-            slice::from_raw_parts(db_data, db_data_len)
-        }));
-
-        WalletDb::for_path(db_data, network)
-            .map(|db| init_wallet_db(&db))
-            .map(|_| 1)
-            .map_err(|e| format_err!("Error while initializing data DB: {}", e))
-    });
-    unwrap_exc_or_null(res)
-}
-
-/// Initialises the data database with the given number of accounts using the given seed.
+/// Sets up the internal structure of the data database.  The value for `seed` may be provided as a
+/// null pointer if the caller wishes to attempt migrations without providing the wallet's seed
+/// value.
+///
+/// Returns 0 if successful, 1 if the seed must be provided in order to execute the requested
+/// migrations, or -1 otherwise.
 ///
-/// Returns the ExtendedSpendingKeys for the accounts. The caller should store these
-/// securely for use while spending.
+/// # Safety
 ///
-/// Call `zcashlc_vec_string_free` on the returned pointer when you are finished with it.
+/// - `db_data` must be non-null and valid for reads for `db_data_len` bytes, and it must have an
+///   alignment of `1`. Its contents must be a string representing a valid system path in the
+///   operating system's preferred representation.
+/// - The memory referenced by `db_data` must not be mutated for the duration of the function call.
+/// - The total size `db_data_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of pointer::offset.
+/// - `seed` must be non-null and valid for reads for `seed_len` bytes, and it must have an
+///   alignment of `1`.
+/// - The memory referenced by `seed` must not be mutated for the duration of the function call.
+/// - The total size `seed_len` must be no larger than `isize::MAX`. See the safety documentation
+///   of pointer::offset.
 #[no_mangle]
-pub extern "C" fn zcashlc_init_accounts_table(
+pub extern "C" fn zcashlc_init_data_database(
```

Comment:
Deferring this to https://github.com/zcash-hackworks/zcash-light-client-ffi/issues/65#issuecomment-1315988668.

Reply:
@str4d has recommended using the existing style, with unsafe blocks in the function body, so that it's easier to identify exactly which parts of the implementation involve unsafe calls. cc/ @daira @sellout

Reply:
Daira informed me that the lint referred to above enables that style to be maintained, but it was only added to stable releases relatively recently (which is why I didn't know about it). In applications, where we just follow latest stable, we should be able to use it.